gpio/src/vhu_gpio.rs

```diff
@@ -0,0 +1,1189 @@
+            Error::DescriptorReadFailed
+        );
+
+        // Invalid out hdr length
+        let flags: Vec<u16> = vec![0, VIRTQ_DESC_F_WRITE];
+        let len: Vec<u32> = vec![100, size_of::<VirtioGpioIrqResponse>() as u32];
+        let desc_chain = prepare_desc_chain_dummy(None, flags, len);
+        assert_eq!(
+            backend
+                .process_events(vec![desc_chain], &vring)
+                .unwrap_err(),
+            Error::UnexpectedDescriptorSize(size_of::<VirtioGpioIrqRequest>(), 100)
+        );
+
+        // Read only in hdr
+        let flags: Vec<u16> = vec![0, 0];
+        let len: Vec<u32> = vec![
+            size_of::<VirtioGpioIrqRequest>() as u32,
+            size_of::<VirtioGpioIrqResponse>() as u32,
+        ];
+        let desc_chain = prepare_desc_chain_dummy(None, flags, len);
+        assert_eq!(
+            backend
+                .process_events(vec![desc_chain], &vring)
+                .unwrap_err(),
+            Error::UnexpectedReadableDescriptor(1)
+        );
+
+        // Invalid in hdr length
+        let flags: Vec<u16> = vec![0, VIRTQ_DESC_F_WRITE];
+        let len: Vec<u32> = vec![size_of::<VirtioGpioIrqRequest>() as u32, 100];
+        let desc_chain = prepare_desc_chain_dummy(None, flags, len);
+        assert_eq!(
+            backend
+                .process_events(vec![desc_chain], &vring)
+                .unwrap_err(),
+            Error::UnexpectedDescriptorSize(size_of::<VirtioGpioIrqResponse>(), 100)
+        );
+
+        // Wait for event without setting irq type first
+        let flags: Vec<u16> = vec![0, VIRTQ_DESC_F_WRITE];
+        let len: Vec<u32> = vec![
+            size_of::<VirtioGpioIrqRequest>() as u32,
+            size_of::<VirtioGpioIrqResponse>() as u32,
+        ];
+        let desc_chain = prepare_desc_chain_dummy(None, flags, len);
+        let desc_chains = vec![desc_chain];
+        backend.process_events(desc_chains.clone(), &vring).unwrap();
+        validate_desc_chains(desc_chains, VIRTIO_GPIO_IRQ_STATUS_INVALID, None);
+
+        // Wait for interrupt with failure msg
+        const GPIO: u16 = 5;
+        // Set irq type
+        let desc_chain = prepare_request_desc_chain(
+            GuestAddress(0),
+            VIRTIO_GPIO_MSG_IRQ_TYPE,
+            GPIO,
+            VIRTIO_GPIO_IRQ_TYPE_EDGE_BOTH as u32,
+            1,
+        );
+        let desc_chains = vec![desc_chain];
+        backend
+            .process_requests(desc_chains.clone(), &vring)
+            .unwrap();
+        validate_desc_chains(desc_chains, VIRTIO_GPIO_STATUS_OK, Some(vec![0]));
+
+        // Wait for interrupt
+        let desc_chain = prepare_event_desc_chain(GuestAddress(0), GPIO);
+        let desc_chains = vec![desc_chain];
+        backend.process_events(desc_chains.clone(), &vring).unwrap();
+
+        while backend.handles.read().unwrap()[GPIO as usize].is_some() {}
+        validate_desc_chains(desc_chains, VIRTIO_GPIO_IRQ_STATUS_INVALID, None);
```

Comment:
It is really hard to understand that the failure is caused by `device.wait_for_irq_result = Err(err);` on line 934.  As a general comment on the #[test] sections, please add enough comments to the test cases so that it is clear so see what is happening.

Reply:
Right. I have added comment for this and checked/improved others as well.